src/riak_kv_qry_buffers.erl

```diff
@@ -420,55 +427,97 @@ do_batch_put(QBufRef, Data, #state{qbufs          = QBufs0,
             end
     end.
 
--spec maybe_add_chunk(#qbuf{}, [data_row()],
-                      non_neg_integer(), non_neg_integer()) ->
-                             {ok, #qbuf{}, non_neg_integer()} | {error, total_qbuf_size_limit_reached | riak_kv_qry_buffers_ldb:errors()}.
-maybe_add_chunk(#qbuf{ldb_ref       = LdbRef,
-                      orig_qry      = ?SQL_SELECT{'ORDER BY' = OrderBy},
-                      chunks_got    = ChunksGot0,
-                      chunks_need   = ChunksNeed,
-                      size          = Size,
-                      total_records = TotalRecords0,
-                      key_field_positions = KeyFieldPositions} = QBuf0,
-                Data,
-                TotalSize, HardWatermark) ->
+-spec add_chunk(#qbuf{}, [data_row()],
+                non_neg_integer(), non_neg_integer(), non_neg_integer()) ->
+                       {ok, #qbuf{}, non_neg_integer()} | {error, total_qbuf_size_limit_reached | riak_kv_qry_buffers_ldb:errors()}.
+add_chunk(#qbuf{ldb_ref       = LdbRefOrDelayedCreateFun,
+                inmem_buffer  = InmemBuffer0,
+                orig_qry      = ?SQL_SELECT{'ORDER BY' = OrderBy},
+                chunks_got    = ChunksGot0,
+                chunks_need   = ChunksNeed,
+                size          = Size,
+                total_records = TotalRecords0,
+                key_field_positions = KeyFieldPositions} = QBuf0,
+          Data,
+          TotalSize, HardWatermark, InmemMax) ->
     ChunkSize = compute_chunk_size(Data),
     case TotalSize + ChunkSize > HardWatermark of
         true ->
             {error, total_qbuf_size_limit_reached};
         false ->
+            %% update stats/counters
+            ChunkId = ChunksGot0,
+            ChunksGot = ChunksGot0 + 1,
             %% ChunkId will be used to construct a new and unique key
             %% for each record. Ideally, this should be the serial
             %% number of the subquery.
-            ChunkId = ChunksGot0,
-            ChunksGot = ChunksGot0 + 1,
-            lager:debug("adding chunk ~b of ~b", [ChunksGot, ChunksNeed]),
+            IsReady = (ChunksNeed == ChunksGot),
```

Comment:
I really don't like "is ready" (both as binding and field name) for this value. `all_chunks_received` would be better, if I'm understanding its role correctly.

I'm also a bit confused why `do_delete_qbuf` refuses to delete an incomplete query buffer. Is there a workflow where premature deletions are avoided this way, or are we just introducing resource leaks?